trieste/models/keras/models.py

```diff
@@ -0,0 +1,293 @@
+            training the model. Defaults to :class:`~trieste.models.optimizer.KerasOptimizer` with
+            :class:`~tf.optimizers.Adam` optimizer, negative log likelihood loss and a dictionary
+            of default arguments for Keras `fit` method: 100 epochs, batch size 100, and verbose 0.
+            See https://keras.io/api/models/model_training_apis/#fit-method for a list of possible
+            arguments.
+        :param bootstrap: Sample with replacement data for training each network in the ensemble.
+            By default set to `False`.
+        :raise ValueError: If ``model`` is not an instance of
+            :class:`~trieste.models.keras.KerasEnsemble` or ensemble has less than two base
+            learners (networks).
+        """
+
+        if not isinstance(model, KerasEnsemble):
+            raise ValueError(
+                f"In DeepEnsemble model must be an instance of a KerasEnsemble, "
+                f"received {type(model)} instead."
+            )
+
+        if optimizer is None:
+            optimizer = KerasOptimizer(tf.keras.optimizers.Adam())
+        self._optimizer = optimizer
+
+        if not isinstance(optimizer.optimizer, tf.keras.optimizers.Optimizer):
+            raise ValueError(
+                f"Optimizer for `DeepEnsemble` model must be an instance of a "
+                f"`tf.keras.optimizers.Optimizer`, "
+                f"received {type(optimizer.optimizer)} instead."
+            )
+
+        super().__init__(optimizer)
+
+        if not self.optimizer.fit_args:
+            self.optimizer.fit_args: Optional[Dict[str, Any]] = {
+                "verbose": 0,
+                "epochs": 100,
+                "batch_size": 100,
+            }
+
+        if self.optimizer.loss is None:
+            self.optimizer.loss = negative_log_likelihood
+
+        model.model.compile(
+            self.optimizer.optimizer,
+            loss=[self.optimizer.loss] * model.ensemble_size,
+            metrics=[self.optimizer.metrics] * model.ensemble_size,
+        )
+
+        self._model = model
+        self._bootstrap = bootstrap
+
+        if self.ensemble_size < 2:
+            raise ValueError(f"Ensemble size must be greater than 1 but got {self.ensemble_size}.")
+
+    def __repr__(self) -> str:
+        """"""
+        return f"DeepEnsemble({self.model!r}, {self.optimizer!r}, {self._bootstrap!r})"
+
+    @property
+    def model(self) -> tf.keras.Model:
+        """ " Returns compiled Keras ensemble model."""
+        return self._model.model
+
+    @property
+    def ensemble_size(self) -> int:
+        """
+        Returns the size of the ensemble, that is, the number of base learners or individual neural
+        network models in the ensemble.
+        """
+        return self._model.ensemble_size
+
+    def prepare_data(
+        self,
+        dataset: Union[Dataset, TensorType],
+    ) -> Union[tuple[Dict[str, TensorType]], Dict[str, TensorType]]:
```

Comment:
why is this one function rather than two? (if nothing else, the typing would be more precise if you split it into `prepare_dataset` and `prepare_query_points`)

Update: I expect this explains some of the type errors you're getting too.

Reply:
I had it like that before but then thought I had a lot of duplicated code...
I was also thinking where should this live, in DeepEnsemble or KerasEnsemble?